syntax = "proto3";

package glossary;

// The GlossaryService manages the lifecycle of glossary terms.
service GlossaryService {
  // Creates a new term. ID is generated by the server.
  rpc AddTerm(AddTermRequest) returns (Term);

  // Retrieves a term by its unique ID.
  rpc GetTerm(GetTermRequest) returns (Term);

  // Retrieves a list of all terms in the glossary.
  rpc GetAllTerms(GetAllTermsRequest) returns (GetAllTermsResponse);

  // Updates an existing term identified by its ID.
  rpc UpdateTerm(UpdateTermRequest) returns (Term);

  // Deletes a term by its unique ID.
  rpc DeleteTerm(DeleteTermRequest) returns (DeleteTermResponse);
}

// Represents a complete glossary term object, typically used for responses.
message Term {
  string id = 1;
  string name = 2;
  string definition = 3;
  string source_url = 4;
}

// Defines the required fields for creating a new term. Does not include an ID.
message AddTermRequest {
  string name = 1;
  string definition = 2;
  string source_url = 3;
}

// Defines the required fields for updating an existing term.
message UpdateTermRequest {
  string id = 1;
  string name = 2;
  string definition = 3;
  string source_url = 4;
}

message GetTermRequest {
  string id = 1;
}

message GetAllTermsRequest {
  // This request is empty as it requires no parameters.
}

message GetAllTermsResponse {
  repeated Term terms = 1;
}

message DeleteTermRequest {
  string id = 1;
}

message DeleteTermResponse {
  bool success = 1;
}